src/methods.jl

```diff
@@ -10,318 +19,446 @@ derivatives.
-        # For high precision on the \ we use Rational, and to prevent overflows we use Int128
-        # At the end we go to Float64 for fast floating point math (rather than rational math)
+# Compute coefficients for the method and cache the result.
+function _coefs(grid::AbstractVector{<:Real}, q::Integer)
+    return get!(_COEFFS_CACHE, (grid, q)) do
+        p = length(grid)
+        # For high precision on the `\`, we use `Rational`, and to prevent overflows we use
+        # `Int128`. At the end we go to `Float64` for fast floating point math, rather than
+        # rational math.
         C = [Rational{Int128}(g^i) for i in 0:(p - 1), g in grid]
         x = zeros(Rational{Int128}, p)
         x[q + 1] = factorial(q)
         return Float64.(C \ x)
     end
 end
 
+# Estimate the bound on the derivative by amplifying the ∞-norm.
+function _make_default_bound_estimator(; condition::Int=DEFAULT_CONDITION)
+    return (f, x) -> condition * maximum(abs, f(x))
+end
 
-# Estimate the bound on the function value and its derivatives at a point
-_estimate_bound(x, cond) = add_tiny(cond * maximum(abs, x))
+function Base.show(io::IO, m::MIME"text/plain", x::FiniteDifferenceMethod)
+    @printf io "FiniteDifferenceMethod:\n"
+    @printf io "  order of method:       %d\n" length(x.grid)
+    @printf io "  order of derivative:   %d\n" x.q
+    @printf io "  grid:                  %s\n" x.grid
+    @printf io "  coefficients:          %s\n" x.coefs
+end
 
 """
-    fdm(m::FiniteDifferenceMethod, f, x[, Val(false)]; kwargs...) -> Real
-    fdm(m::FiniteDifferenceMethod, f, x, Val(true); kwargs...) -> Tuple{FiniteDifferenceMethod, Real}
-
-Compute the derivative of `f` at `x` using the finite differencing method `m`.
-The optional `Val` argument dictates whether the method should be returned alongside the
-derivative value, which can be useful for examining the step size used and other such
-parameters.
-
-The recognized keywords are:
+    function estimate_step(
+        m::FiniteDifferenceMethod,
+        f::Function,
+        x::T;
+        factor::Int=1,
+        max_step=convert(T, 0.1)
+    ) where T<:AbstractFloat
+
+Estimate the step size for a finite difference method `m`. Also estimates the error of the
+estimate of the derivative.
+
+# Arguments
+- `m::FiniteDifferenceMethod`: Finite difference method to estimate the step size for.
+- `f::Function`: Function to evaluate the derivative of.
+- `x::T`: Point to estimate the derivative at.
+
+# Keywords
+- `factor::Int=1`. Factor to amplify the estimated round-off error by. This can be used
+    to force a more conservative step size.
+- `max_step=convert(T, 0.1)`: Maximum step size.
+
+# Returns
+- `Tuple{T, T}`: Estimated step size and an estimate of the error of the finite difference
+    estimate.
+"""
+function estimate_step(
+    m::FiniteDifferenceMethod,
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
+) where T<:AbstractFloat
```

Comment:
```suggestion
)::Tuple{T, T} where T<:AbstractFloat
```
I did some quick investigation, and this is currently failing to type-infer.
Possibly its not type-stable.
But anyway adding this convert-constraint gives a 3x performance vs
before
https://github.com/JuliaDiff/FiniteDifferences.jl/pull/109#issuecomment-701342044


```
julia> @btime $_fdm(sin, 0.2);
  1.393 μs (34 allocations: 1.66 KiB)
```

Reply:
This would assumes that if the input of `f` is of type `T` then the output is also of type `T`. Is that an assumption we're willing to make?

Reply:
Hmm probably not, but it is what the docstring says it does